File: 62 Unique Paths/Solution.js
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths = function(m, n) {
    res = new Array(n);
    res[0] = 1;
    
    for (i = 0; i < m; i++) {
        for (j = 1; j < n; j++) {
            if (isNaN(res[j]))
                res[j] = 0;
            res[j] += res[j-1];
        }
    }
    
    return res[n-1];
};

File: 62 Unique Paths/Solution.java
public class Solution {
    public int uniquePaths(int m, int n) {
        int[][] res = new int[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                res[i][j] = 1;
            }
        }
        
        //都从1开始
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                res[i][j] = res[i-1][j] + res[i][j-1];
            }
        }
        
        return res[m-1][n-1];
    }
}


//省内存版
public class Solution {
    public int uniquePaths(int m, int n) {
        int[] res = new int[n];
        res[0] = 1;
        
        for (int i = 0; i < m; i++){
            for (int j = 1; j < n; j++) {
                res[j] += res[j-1];
            }
        }
        
        return res[n-1]
    }
}